contracts/Staking.sol

```diff
@@ -113,148 +96,189 @@ contract Staking is Autopetrified, ERCStaking, ERCStakingHistory, IStakingLockin
      * @param _accountAddress Owner of locked tokens
-     * @param _lockId Id of the lock for the given account
-     * @param _newAmount New amount of locked tokens
+     * @param _lockManager The manager entity for this particular lock
+     * @param _allowance Amount of allowed tokens decrease
      */
-    function decreaseLockAmount(address _accountAddress, uint256 _lockId, uint256 _newAmount) external isLockManager(_accountAddress, _lockId) {
-        // lock 0 tokens makes no sense
-        require(_newAmount > 0, ERROR_AMOUNT_ZERO);
+    function decreaseLockAllowance(address _accountAddress, address _lockManager, uint256 _allowance) external isInitialized {
+        // only owner and manager can decrease allowance
+        require(msg.sender == _accountAddress || msg.sender == _lockManager, ERROR_CANNOT_CHANGE_ALLOWANCE);
+        require(_allowance > 0, ERROR_AMOUNT_ZERO);
 
-        // manager can only decrease locked amount
-        Lock storage lock_ = accounts[_accountAddress].locks[_lockId];
-        require(_newAmount < lock_.amount, ERROR_INCREASING_LOCK_AMOUNT);
+        Lock storage lock = accounts[_accountAddress].locks[_lockManager];
+        uint256 newAllowance = lock.allowance.sub(_allowance);
+        require(newAllowance >= lock.amount, ERROR_NOT_ENOUGH_ALLOWANCE);
+        // unlock must be used for this:
+        require(newAllowance > 0, ERROR_ALLOWANCE_ZERO);
 
-        lock_.amount = _newAmount;
-        emit LockAmountChanged(_accountAddress, _lockId, _newAmount);
+        lock.allowance = newAllowance;
+
+        emit LockAllowanceChanged(_accountAddress, _lockManager, _allowance, false);
     }
 
     /**
-     * @notice Change the manager of `_accountAddress`'s lock #`_lockId` to `_newManager`
-     * @param _accountAddress Owner of lock
-     * @param _lockId Id of the lock for the given account
-     * @param _newManager New lock's manager
+     * @notice Increase locked amount by `@tokenAmount(stakingToken: address, _amount)` for user `_accountAddress` by lock manager `_lockManager`
+     * @param _accountAddress Owner of locked tokens
+     * @param _lockManager The manager entity for this particular lock
+     * @param _amount Amount of locked tokens increase
      */
-    function setLockManager(
-        address _accountAddress,
-        uint256 _lockId,
-        ILockManager _newManager
-    )
-        external
-        isLockManager(_accountAddress, _lockId)
-    {
-        accounts[_accountAddress].locks[_lockId].manager = _newManager;
-        emit LockManagerChanged(_accountAddress, _lockId, _newManager);
+    function increaseLockAmount(address _accountAddress, address _lockManager, uint256 _amount) external isInitialized {
+        require(_amount > 0, ERROR_AMOUNT_ZERO);
+
+        // check enough unlocked tokens are available
+        require(_amount <= _unlockedBalanceOf(_accountAddress), ERROR_NOT_ENOUGH_BALANCE);
+
+        _increaseActiveLockAmount(_accountAddress, _lockManager, _amount);
```

Comment:
🚨Shouldn't we validate the sender somewhere here? 

If that's the case, let's identify what was the case missing and write/fix a test for this

Reply:
Definitely, I think it used to be called with `msg.sender` and I messed it up refactoring. Will fix it and add tests for it.